File: src/go/svd/models/deviation_from_planned.go
package models

import "time"

// DeviationFromPlanned represents the "Deviation From Planned" section of SVD.
type DeviationFromPlanned struct {
	Reason                   string    `json:"reason" xml:"Reason" csv:"Reason"`
	Latitude                 float64   `json:"latitude" xml:"Latitude" csv:"Latitude"`
	Longitude                float64   `json:"longitude" xml:"Longitude" csv:"Longitude"`
	ShipDeviationStartedTime time.Time `json:"shipDeviationStartedTime" xml:"ShipDeviationStartedTime" csv:"ShipDeviationStartedTime"`
	ShipDeviationStoppedTime time.Time `json:"shipDeviationStoppedTime" xml:"ShipDeviationStoppedTime" csv:"ShipDeviationStoppedTime"`
}

File: src/elixir/svd/lib/svd/models/deviation_from_planned.ex
defmodule SVD.Models.DeviationFromPlanned do
  @moduledoc false

  @fields [
    :deviation_reason,
    :deviation_description,
    :deviation_distance,
    :deviation_time,
    :reason,
    :latitude,
    :longitude,
    :ship_deviation_started_time,
    :ship_deviation_stopped_time
  ]

  @derive {Jason.Encoder, only: @fields}
  defstruct @fields

  @type t :: %__MODULE__{}

  @spec fields() :: [atom()]
  def fields, do: @fields
end
